File: backend/models/Medicine.js
/**
 * Medicine.js — CSV-based medicine search (no Supabase)
 * 
 * Loads the A_Z_medicines_dataset_of_India.csv into memory once on startup.
 * Provides fast in-memory search across 250k+ medicines.
 *
 * CSV columns:
 *   id, name, price(₹), Is_discontinued, manufacturer_name,
 *   type, pack_size_label, short_composition1, short_composition2
 */

const fs   = require('fs');
const path = require('path');
const readline = require('readline');

const CSV_PATH = path.join(__dirname, '../data/medicines.csv');

// In-memory store
let medicines = [];
let loaded    = false;
let loading   = false;
let loadError = null;

// ── Parse a CSV line respecting quoted fields ────────────────
function parseCSVLine(line) {
  const fields = [];
  let current  = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      inQuotes = !inQuotes;
    } else if (ch === ',' && !inQuotes) {
      fields.push(current.trim());
      current = '';
    } else {
      current += ch;
    }
  }
  fields.push(current.trim());
  return fields;
}

// ── Load CSV into memory ─────────────────────────────────────
function loadCSV() {
  return new Promise((resolve, reject) => {
    if (loaded)  return resolve(medicines);
    if (loading) {
      // Wait for existing load to finish
      const wait = () => {
        if (loaded)    return resolve(medicines);
        if (loadError) return reject(loadError);
        setTimeout(wait, 100);
      };
      return wait();
    }

    loading = true;
    console.log('[Medicine] Loading CSV from disk…');

    const rl = readline.createInterface({
      input:     fs.createReadStream(CSV_PATH, { encoding: 'utf8' }),
      crlfDelay: Infinity,
    });

    let header  = null;
    let rowCount = 0;

    rl.on('line', (line) => {
      if (!line.trim()) return;

      const fields = parseCSVLine(line);

      if (!header) {
        // First line is header — normalise column names
        header = fields.map(h =>
          h.replace(/\(.*?\)/g, '').replace(/[^a-zA-Z0-9_]/g, '_').toLowerCase()
        );
        return;
      }

      const raw = {};
      header.forEach((col, i) => { raw[col] = fields[i] || ''; });

      // Map to a clean object
      const med = {
        id:               parseInt(raw.id) || rowCount,
        name:             raw.name || '',
        price:            parseFloat(raw.price) || 0,
        market_price:     parseFloat(raw.price) || 0,
        is_discontinued:  (raw.is_discontinued || '').toLowerCase() === 'true',
        manufacturer:     raw.manufacturer_name || '',
        manufacturer_name:raw.manufacturer_name || '',
        type:             raw.type || 'allopathy',
        category:         raw.type || 'allopathy',
        pack_size:        raw.pack_size_label || '',
        pack_size_label:  raw.pack_size_label || '',
        // Combine both compositions as generic_name
        generic_name: [raw.short_composition1, raw.short_composition2]
          .filter(Boolean).join(' + ').trim(),
        short_composition1: raw.short_composition1 || '',
        short_composition2: raw.short_composition2 || '',
      };

      if (med.name) {
        medicines.push(med);
        rowCount++;
      }
    });

    rl.on('close', () => {
      loaded  = true;
      loading = false;
      console.log(`[Medicine] ✅ Loaded ${medicines.length.toLocaleString()} medicines from CSV`);
      resolve(medicines);
    });

    rl.on('error', (err) => {
      loadError = err;
      loading   = false;
      console.error('[Medicine] ❌ CSV load error:', err.message);
      reject(err);
    });
  });
}

// ── Start loading in background when module is required ──────
loadCSV().catch(err => console.error('[Medicine] Background load failed:', err.message));

// ── Helpers ──────────────────────────────────────────────────
function normalize(str) {
  return (str || '').toLowerCase().trim();
}

function compact(str) {
  return (str || '').toLowerCase().replace(/[^a-z0-9]/g, '');
}

function matchesQuery(med, q) {
  const lq = normalize(q);
  const cq = compact(q);
  const cn = compact(med.name);
  const cg = compact(med.generic_name);
  return (
    normalize(med.name).includes(lq) ||
    normalize(med.generic_name).includes(lq) ||
    (cq.length >= 2 && (cn.includes(cq) || cg.includes(cq))) ||
    normalize(med.manufacturer).includes(lq) ||
    normalize(med.category).includes(lq) ||
    normalize(med.short_composition1).includes(lq) ||
    normalize(med.short_composition2).includes(lq)
  );
}

// ── Public API (mirrors old Supabase model) ──────────────────
const Medicine = {

  async search(query, limit = 30) {
    await loadCSV();
    const lq = normalize(query);

    // Prioritise name-starts-with matches, then contains
    const startsWith = [];
    const contains   = [];

    for (const med of medicines) {
      if (!med.name) continue;
      const nameLower = normalize(med.name);

      if (nameLower.startsWith(lq)) {
        startsWith.push(med);
      } else if (matchesQuery(med, lq)) {
        contains.push(med);
      }

      if (startsWith.length + contains.length >= limit * 3) break;
    }

    return [...startsWith, ...contains].slice(0, limit);
  },

  // Paginated search — returns { total, medicines }
  async searchPaginated(query, page = 1, limit = 24) {
    await loadCSV();
    const lq = normalize(query);
    const startsWith = [];
    const contains   = [];

    for (const med of medicines) {
      if (!med.name) continue;
      const nameLower = normalize(med.name);
      if (nameLower.startsWith(lq))    startsWith.push(med);
      else if (matchesQuery(med, lq))  contains.push(med);
    }

    const all    = [...startsWith, ...contains];
    const total  = all.length;
    const offset = (page - 1) * limit;
    return { total, medicines: all.slice(offset, offset + limit) };
  },

  // Browse all medicines by first letter, sorted A-Z, paginated
  async browse(letter = 'A', page = 1, limit = 24) {
    await loadCSV();
    const ltr = (letter || 'A').toUpperCase();

    const filtered = medicines
      .filter(m => m.name && m.name.toUpperCase().startsWith(ltr))
      .sort((a, b) => a.name.localeCompare(b.name));

    const total  = filtered.length;
    const offset = (page - 1) * limit;
    return { total, medicines: filtered.slice(offset, offset + limit) };
  },

  async findByName(name) {
    await loadCSV();
    const lq = normalize(name);
    return medicines.find(m => normalize(m.name) === lq)
        || medicines.find(m => normalize(m.name).includes(lq))
        || null;
  },

  async findByCategory(category, limit = 50) {
    await loadCSV();
    const lq = normalize(category);
    return medicines
      .filter(m => normalize(m.category).includes(lq))
      .slice(0, limit);
  },

  async findAlternatives(medicineName, limit = 10) {
    await loadCSV();
    const original = await Medicine.findByName(medicineName);
    if (!original || !original.generic_name) return [];

    const mainGeneric = normalize(original.generic_name.split('+')[0].split('(')[0]);

    return medicines
      .filter(m => m.id !== original.id && normalize(m.generic_name).includes(mainGeneric))
      .sort((a, b) => a.price - b.price)
      .slice(0, limit);
  },

  async getCategories() {
    await loadCSV();
    const counts = {};
    for (const m of medicines) {
      const cat = m.category || 'Other';
      counts[cat] = (counts[cat] || 0) + 1;
    }
    return Object.entries(counts)
      .map(([name, count]) => ({ name, count }))
      .sort((a, b) => b.count - a.count);
  },

  async count() {
    await loadCSV();
    return medicines.length;
  },

  // Expose raw for diagnostics
  isLoaded() { return loaded; },
};

module.exports = Medicine;

File: backend/models/medicine.py
"""
medicine.py — In-memory medicine search across 250k+ medicines from data/medicines.csv
Mirrors backend/models/Medicine.js
"""
import csv
from pathlib import Path
from typing import List, Dict, Any, Optional

DATA_PATH = Path(__file__).parent.parent / "data" / "medicines.csv"

_medicines: List[Dict[str, Any]] = []
_loaded: bool = False


def _normalize(s: Optional[str]) -> str:
    return (s or "").lower().strip()


def load_medicines() -> List[Dict[str, Any]]:
    global _medicines, _loaded
    if _loaded:
        return _medicines

    if not DATA_PATH.exists():
        return []

    items = []
    try:
        with open(DATA_PATH, mode="r", encoding="utf-8", errors="replace") as f:
            reader = csv.DictReader(f)
            # normalize field names
            for idx, row in enumerate(reader):
                # Clean headers
                clean_row = {
                    k.replace("(", "").replace(")", "").replace("₹", "").strip().lower(): v
                    for k, v in row.items() if k
                }
                name = clean_row.get("name", "").strip()
                if not name:
                    continue

                price_val = 0.0
                raw_price = clean_row.get("price", "0")
                try:
                    price_val = float(raw_price)
                except ValueError:
                    price_val = 0.0

                comp1 = clean_row.get("short_composition1", "").strip()
                comp2 = clean_row.get("short_composition2", "").strip()
                generic_parts = [c for c in [comp1, comp2] if c]
                generic_name = " + ".join(generic_parts)

                m = {
                    "id": int(clean_row.get("id", idx)) if clean_row.get("id", "").isdigit() else idx,
                    "name": name,
                    "price": price_val,
                    "market_price": price_val,
                    "is_discontinued": clean_row.get("is_discontinued", "").lower() == "true",
                    "manufacturer": clean_row.get("manufacturer_name", ""),
                    "manufacturer_name": clean_row.get("manufacturer_name", ""),
                    "type": clean_row.get("type", "allopathy"),
                    "category": clean_row.get("type", "allopathy"),
                    "pack_size": clean_row.get("pack_size_label", ""),
                    "pack_size_label": clean_row.get("pack_size_label", ""),
                    "generic_name": generic_name,
                    "short_composition1": comp1,
                    "short_composition2": comp2,
                }
                items.append(m)
        _medicines = items
        _loaded = True
        print(f"[Medicine] Loaded {len(_medicines):,} medicines from CSV")
    except Exception as e:
        print(f"[Medicine] Error loading CSV: {e}")

    return _medicines


def _matches_query(med: Dict[str, Any], query_lower: str) -> bool:
    return (
        query_lower in _normalize(med["name"]) or
        query_lower in _normalize(med["generic_name"]) or
        query_lower in _normalize(med["manufacturer"]) or
        query_lower in _normalize(med["category"]) or
        query_lower in _normalize(med["short_composition1"]) or
        query_lower in _normalize(med["short_composition2"])
    )


class Medicine:
    @classmethod
    def search(cls, query: str, limit: int = 30) -> List[Dict[str, Any]]:
        meds = load_medicines()
        lq = _normalize(query)
        starts_with = []
        contains = []

        for med in meds:
            nl = _normalize(med["name"])
            if nl.startswith(lq):
                starts_with.append(med)
            elif _matches_query(med, lq):
                contains.append(med)

            if len(starts_with) + len(contains) >= limit * 3:
                break

        return (starts_with + contains)[:limit]

    @classmethod
    def search_paginated(cls, query: str, page: int = 1, limit: int = 24) -> Dict[str, Any]:
        meds = load_medicines()
        lq = _normalize(query)
        starts_with = []
        contains = []

        for med in meds:
            nl = _normalize(med["name"])
            if nl.startswith(lq):
                starts_with.append(med)
            elif _matches_query(med, lq):
                contains.append(med)

        all_matches = starts_with + contains
        total = len(all_matches)
        offset = (page - 1) * limit
        return {
            "total": total,
            "medicines": all_matches[offset: offset + limit]
        }

    @classmethod
    def browse(cls, letter: str = "A", page: int = 1, limit: int = 24) -> Dict[str, Any]:
        meds = load_medicines()
        ltr = (letter or "A").upper()

        filtered = [m for m in meds if m["name"] and m["name"].upper().startswith(ltr)]
        filtered.sort(key=lambda x: x["name"])

        total = len(filtered)
        offset = (page - 1) * limit
        return {
            "total": total,
            "medicines": filtered[offset: offset + limit]
        }

    @classmethod
    def find_by_name(cls, name: str) -> Optional[Dict[str, Any]]:
        meds = load_medicines()
        lq = _normalize(name)
        for m in meds:
            if _normalize(m["name"]) == lq:
                return m
        for m in meds:
            if lq in _normalize(m["name"]):
                return m
        return None

    @classmethod
    def find_by_category(cls, category: str, limit: int = 50) -> List[Dict[str, Any]]:
        meds = load_medicines()
        lq = _normalize(category)
        return [m for m in meds if lq in _normalize(m["category"])][:limit]

    @classmethod
    def find_alternatives(cls, medicine_name: str, limit: int = 10) -> List[Dict[str, Any]]:
        meds = load_medicines()
        original = cls.find_by_name(medicine_name)
        if not original or not original.get("generic_name"):
            return []

        main_generic = _normalize(original["generic_name"].split("+")[0].split("(")[0])
        matches = [
            m for m in meds
            if m["id"] != original["id"] and main_generic in _normalize(m["generic_name"])
        ]
        matches.sort(key=lambda x: x.get("price", 0.0))
        return matches[:limit]

    @classmethod
    def get_categories(cls) -> List[Dict[str, Any]]:
        meds = load_medicines()
        counts: Dict[str, int] = {}
        for m in meds:
            cat = m.get("category") or "Other"
            counts[cat] = counts.get(cat, 0) + 1
        sorted_counts = sorted(counts.items(), key=lambda x: x[1], reverse=True)
        return [{"name": name, "count": count} for name, count in sorted_counts]

    @classmethod
    def count(cls) -> int:
        return len(load_medicines())

    @classmethod
    def is_loaded(cls) -> bool:
        return _loaded
